### ISSUE TITLE
geth crashes frequently when trying to sync the blockchain

### ISSUE THREAD

**Issue by blackout24** [ethereum/go-ethereum#1437]
Hello,

on Arch Linux (x86_64) with Go 1.4.2 geth crashes very frequently when I do a fresh sync to a data dir on my SSD. 

I cloned the repo and ran "make all" on the latest commit on the develop branch yesterday. It started to initialize a data dir in my home directory and block chain synchronization started. After about 30 minutes geth crashed. Thinking that the latest development snapshot might be unstable I  git checkout the v0.9.36 tag. Ran make clean and make all and restarted geth. It started syncing but it crashed again after a while. I restarted numerous times, but never managed to sync for very long. Sometimes it just crashed after it imported the first 256 blocks. I also have the feeling that the crashes are more frequent from block 300K on. No I'm at ~600K and can't run geth for more than a minute.

Here is the crashlog:
https://gist.github.com/blackout24/26c7ee561047f8e070b7

This is the crashlog after that it only managed to run for 9 minutes:
https://gist.github.com/blackout24/6492e3d186c823b4449f

**Comment by karalabe** [comment]
Whoah... that's (seems at least) an upstream LevelDB bug :|... will try and report it and see what they say.

**Comment by karalabe** [comment]
Could you please try running with this PR? https://github.com/ethereum/go-ethereum/pull/1446

If the problem still persists (I'm guessing it should), could you please build and run geth with the race detector enabled? From the project root:

```
godep go build --race ./cmd/geth
./geth <your usual parameters> (note, the ./ to run the fresh build)
```

If you don't have godeps, you can do a `go get github.com/tools/godep` to install it.

**Comment by blackout24** [comment]
So I git cloned your repo into a different directory checkout the fix-database-metric-race tag and ran "make geth" and started it so it picked up my existing data directory. It still crashed after 10 minutes.
https://gist.github.com/blackout24/9d02315b236f7fb26ac9

So I built it with the race detector.
https://gist.github.com/blackout24/7d4d4b232049b0d4fd95

Kind of wondering why I'm the only one that has this problem. For me this is so easy to reproduce, which makes me wonder if I'm a special corner case.

**Comment by karalabe** [comment]
From the logs, it seems to me you ran out of memory. How much does your system have, and do you have swap enabled or not?

**Comment by karalabe** [comment]
Maybe not, can you also please send a detailed version of your kernel?

**Comment by karalabe** [comment]
After reading a bit after it, it seems the Arch Linux Go distribution is broken (https://github.com/golang/go/issues/6787#issuecomment-101838714). Are you using the default one provided by the package manager, or manual install from the golang site/sources? If you're using the package manager, could you try uninstalling that and setting up manually from http://golang.org/doc/install ?

**Comment by blackout24** [comment]
Hi,

this is interesting I have always wondered why some of my swap space filed up, even though I never noticed that I had dramatically increased RAM usage while running geth.
My PC has 8 GB of RAM and 2GB of ZRAM SWAP. When I start geth and let it sync the RAM usage is just under 2 GB and doesn't really grow, but I noticed that there are 500 MB in swap probably from this out of memory condition. With the stuff I usually do with my PC swap basically is never used. I have vm.swappiness = 60 which is the default.

I use the stock Arch kernel config:
https://projects.archlinux.org/svntogit/packages.git/tree/trunk/config.x86_64?h=packages/linux
I'm using go installed via the package manager. I'll try a manual installation.

**Comment by blackout24** [comment]
Since Arch builds the go package from source according to the PKGBUILD I downloaded the tarball distribution. Before that I removed the go package. I extracted the tarball in my Download directory and exported the new PATH and GOROOT variables. I cleaned my existing go-ethereum build and ran "make geth". I'm now using the v0.9.38 tag.

It still crashed. 
https://gist.github.com/blackout24/10f991748ee5186248f1 
I also monitored my RAM usage with the Gnome System Monitor while it crashed, but I didn't even see a spike in RAM usage when it crashed. Maybe it happens so fast that the system monitor interval (1 s) doesn't pick it up.

I enabled the race detector and the log seems to be the same as the old one:
https://gist.github.com/blackout24/53a0ca82f825ce5d7da7

**Comment by karalabe** [comment]
Please try the official Go distribution from the Go website https://golang.org/dl/. According to the post I linked in my previous comment the Arch package is misconfigured, so I would guess both binaries and sources in the Arch repos are bad.

**Comment by blackout24** [comment]
I did use the official prebuilt tarball distribution when it crashed in the reports above. 
I'm monitoring the resource usage now after boot up. I use 800 MB on an empty desktop. When I run geth it slowly climbs to 1.6 GB used and now seems pretty stable at that number. However my system is constantly swapping small amounts. It's now up to 300 MB already after maybe an hour.  

I think this might be related to the IO and and the VFS in Linux that might be doing some caching. This would explain, why I did not have problems syncing to my HDD which as probably just to slow to require that. This would also explain why crashes came ever more frequent, because swap never was empty after geth crashed so with the next try it filled up faster.
